nncf/experimental/tensor/functions/dispatcher.py

```diff
@@ -8,50 +8,150 @@
-from typing import List
-
-import numpy as np
+from functools import _find_impl
+from inspect import getfullargspec
+from inspect import isclass
+from inspect import isfunction
+from types import MappingProxyType
+from typing import List, get_type_hints
 
 from nncf.experimental.tensor import Tensor
 
 
-def tensor_guard(func: callable):
+def _get_target_types(type_alias):
+    if isclass(type_alias):
+        return [type_alias]
+    ret = []
+    for t in type_alias.__args__:
+        ret.extend(_get_target_types(t))
+    return ret
+
+
+def tensor_dispatch(func):
     """
-    A decorator that ensures that the first argument to the decorated function is a Tensor.
+    This decorator creates a registry of functions for different types and provides a wrapper
+    that calls the appropriate function based on the type of the first argument.
+    It's particularly designed to handle Tensor inputs and outputs effectively.
+
+    :param func: The function to be decorated.
+    :return: The decorated function with type-based dispatching functionality.
     """
 
-    @functools.wraps(func)
-    def wrapper(*args, **kwargs):
-        if isinstance(args[0], Tensor):
-            return func(*args, **kwargs)
-        raise NotImplementedError(f"Function `{func.__name__}` is not implemented for {type(args[0])}")
+    registry = {}
 
-    return wrapper
+    def dispatch(cls):
+        """
+        Retrieves the registered function for a given type.
 
+        :param cls: The type to retrieve the function for.
+        :return: The registered function for the given type.
+        """
+        try:
+            return registry[cls]
+        except KeyError:
+            return _find_impl(cls, registry)
 
-def dispatch_list(fn: "functools._SingleDispatchCallable", tensor_list: List[Tensor], *args, **kwargs):
-    """
-    Dispatches the function to the type of the wrapped data of the first element in tensor_list.
+    def register(rfunc):
+        """Registers a function for a specific type or types.
 
-    :param fn: A function wrapped by `functools.singledispatch`.
-    :param tensor_list: List of Tensors.
-    :return: The result value of the function call.
-    """
-    unwrapped_list = [i.data for i in tensor_list]
-    return fn.dispatch(type(unwrapped_list[0]))(unwrapped_list, *args, **kwargs)
+        :param rfunc: The function to register.
+        :return: The registered function.
+        """
+        assert isfunction(rfunc), "Register object should be a function."
+        assert getfullargspec(func)[0] == getfullargspec(rfunc)[0], "Differ names of arguments of function"
 
+        target_type_hint = get_type_hints(rfunc).get(getfullargspec(rfunc)[0][0])
+        assert target_type_hint is not None, "No type hint for first argument of function"
 
-def register_numpy_types(singledispatch_fn):
-    """
-    Decorator to register function to singledispatch for numpy classes.
+        types_to_registry = set(_get_target_types(target_type_hint))
```

Comment:
Can we move this function from the `tensor_dispatcher` function body? So we won't create a separate set of functions for each `tensor_dispatcher` call and will same a bit of memory